components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCPermissionBasedInternalScopeValidator.java

```diff
@@ -192,6 +233,90 @@
+     * federated users.
+     * @param authenticatedUser    FederatedAuthenticatedUser
+     * @param authorizationManager AuthorizationManager
+     * @return List of permissions
+     * @throws UserStoreException      UserStoreException
+     * @throws IdentityOAuth2Exception IdentityOAuth2Exception
+     */
+    private String[] getAllowedUIResourcesForNotAssociatedFederatedUser(AuthenticatedUser authenticatedUser,
+                                                                        AuthorizationManager authorizationManager)
+            throws UserStoreException, IdentityOAuth2Exception {
+
+        List<String> userRolesList = new ArrayList<>();
+        List<String> allowedUIResourcesListForUser = new ArrayList<>();
+        IdentityProvider identityProvider =
+                OAuth2Util.getIdentityProvider(authenticatedUser.getFederatedIdPName(),
+                        authenticatedUser.getTenantDomain());
+        /*
+        Values of Groups consists mapped local roles and Internal/everyone corresponding to
+        authenticated user.
+        Role mapping consists mapped federated roles with local roles corresponding to IDP.
+        By cross checking role mapped local roles and values of groups we can filter valid local roles which mapped
+        to a federated role of authenticated user.
+         */
+        List<String> valuesOfGroups = getValuesOfGroupsFromUserAttributes(authenticatedUser.getUserAttributes());
+        if (CollectionUtils.isNotEmpty(valuesOfGroups)) {
+            for (RoleMapping roleMapping : identityProvider.getPermissionAndRoleConfig().getRoleMappings()) {
+                if (roleMapping != null && roleMapping.getLocalRole() != null) {
+                    if (valuesOfGroups.contains(roleMapping.getLocalRole().getLocalRoleName())) {
+                        userRolesList.add(roleMapping.getLocalRole().getLocalRoleName());
+                    }
+                }
+            }
+        }
+        // Loop through each local role and get permissions.
+        for (String userRole : userRolesList) {
+            for (String allowedUIResource : authorizationManager.getAllowedUIResourcesForRole(userRole, "/")) {
+                if (!allowedUIResourcesListForUser.contains(allowedUIResource)) {
+                    allowedUIResourcesListForUser.add(allowedUIResource);
+                }
+            }
+        }
+        // Add everyone permission to allowed permission.
+        allowedUIResourcesListForUser.add(EVERYONE_PERMISSION);
+
+        return allowedUIResourcesListForUser.toArray(new String[0]);
+    }
+
+    /**
+     * Get groups params Roles from User attributes.
+     *
+     * @param userAttributes User Attributes
+     * @return User attribute map
+     */
+    private List<String> getValuesOfGroupsFromUserAttributes(Map<ClaimMapping, String> userAttributes) {
+
+        if (isNotEmpty(userAttributes)) {
```

Comment:
Shall we use the full qualified method here instead of importing MapUtils.isNotEmpty()?